src/orange/surf/detail/FerrariSolver.hh

```diff
@@ -0,0 +1,347 @@
+ * consequence of the particular algorithm chosen. This solver implements the
+ * Ferrari-Cardano method, which is well-established and simple, but more
+ * prone to numerical error than contemporary methods to be explored such as
+ * Algorithm 1010\citet{orellana-alg1010-2020,
+ * https://doi.org/10.1145/3386241}.
+ *
+ * \return An Intersections array where each item is either a positive valid
+ * intersection or the sentinel result \c no_intersection().
+ */
+class FerrariSolver
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Intersections = Array<real_type, 4>;
+    using Real2 = Array<real_type, 2>;
+    using Real5 = Array<real_type, 5>;
+    //!@}
+
+    // Solve with manually specified surface state
+    static inline CELER_FUNCTION Intersections solve_general(
+        Real5 abcde,
+        SurfaceState on_surface,
+        real_type tolerance = Tolerance<real_type>::sqrt_quadratic());
+
+  public:
+    // Construct w/ given tolerance. Uses quadratic tolerance by default.
+    inline CELER_FUNCTION
+    FerrariSolver(real_type tolerance = Tolerance<real_type>::sqrt_quadratic());
+
+    // Solver for fully general case
+    inline CELER_FUNCTION Intersections operator()(Real5 abcde) const;
+
+  private:
+    //// DATA ////
+    // Soft zero for biquadratic and degenerate cubic detection
+    static inline SoftZero<real_type> const soft_zero_;
+
+    //// UTIL ////
+    // Try to place real at given index in list, return next free index
+    static inline CELER_FUNCTION int
+    place_root(Intersections& roots, real_type new_root, int free_index);
+
+    // Find roots of special reduced quartic which is biquadratic
+    static inline CELER_FUNCTION Intersections
+    calc_biquadratic_roots(real_type qb, real_type p, real_type r);
+
+    // Find dominant root of normalized cubic
+    static inline CELER_FUNCTION real_type
+    dominant_root_normalized_cubic(real_type b, real_type c, real_type d);
+
+    // Find real quadratic roots
+    static inline CELER_FUNCTION Real2
+    real_roots_normalized_quadratic(real_type b, real_type c);
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Find all positive roots for quartic surfaces using Ferrari-Cardano method.
+ *
+ * This method allows the user to manually specify if the particle starts on
+ * the surface, and use 0 instead of e for accuracy.
+ */
+CELER_FUNCTION auto FerrariSolver::solve_general(Real5 abcde,
+                                                 SurfaceState on_surface,
+                                                 real_type tolerance)
+    -> Intersections
+{
+    FerrariSolver solve(tolerance);
+    auto selected_abcde = abcde;
+    if (on_surface == SurfaceState::on)
+    {
+        auto [a, b, c, d, e] = abcde;
+        selected_abcde = {a, b, c, d, 0};
+    }
+    return solve(selected_abcde);
```

Comment:
Since `abcde` is a local variable you can simply do
```suggestion
    if (on_surface == SurfaceState::on)
    {
        abcde[4] = 0;
    }
    return solve(abcde);
```